pkg/virt-handler/rest/console.go

```diff
@@ -138,20 +138,40 @@ func (t *ConsoleHandler) USBRedirHandler(request *restful.Request, response *res
 	t.stream(vmi, request, response, unixSocketDialer(vmi, unixSocketPath), stopChan)
 }
 
+func (t *ConsoleHandler) vncInUse(uid types.UID) bool {
+	t.vncLock.Lock()
+	defer t.vncLock.Unlock()
+	_, inUse := t.vncStopChans[uid]
+	return inUse
+}
+
 func (t *ConsoleHandler) VNCHandler(request *restful.Request, response *restful.Response) {
 	vmi, code, err := getVMI(request, t.vmiStore)
 	if err != nil {
 		log.Log.Object(vmi).Reason(err).Error(failedRetrieveVMI)
 		response.WriteError(code, err)
 		return
 	}
+
+	uid := vmi.GetUID()
+	preserveSessionParam, err := strconv.ParseBool(request.QueryParameter("preserveSession"))
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Warningf("VNC's query parameter preserveSession")
+	}
+
+	if t.vncInUse(uid) && preserveSessionParam {
```

Comment:
nit: I would swap the order of conditions here, because of short-circuit evaluation. The function that takes a lock would only be called if the `preserveSessionParam` is `true`.